### PR TITLE
ELF Loading in User Space

### PR DESCRIPTION
**Pull request by dayeol**
This makes ELF loading be completely done in the user space.
The ELF loader also parses the entry point of each ELFs.
Following functions were deprecated.
```
Params::setEnclaveEntry()
Params::getEnclaveEntry()
Params::setRuntimeEntry()
Params::getRuntimeEntry()
```

### CODE REVIEW

**Comment by dayeol** [comment]
Since this is a huge PR, I recommend to look directly into the code in order to review, instead of using web UI. The files you may want to closely review are `lib/host/src/keystone.cpp`, `lib/host/src/elffile.cpp`.

File: lib/host/src/keystone.cpp [hunk 1]
```diff
@@ -4,111 +4,275 @@
...
+
+  ret = ioctl(fd, KEYSTONE_IOC_FINALIZE_ENCLAVE, &enclp);
+
+  if (ret) {
+    ERROR("failed to finalize enclave - ioctl() failed: %d", ret);
+    return KEYSTONE_ERROR;
+  }
+
+  /* ELF files are no longer needed */
+  delete enclaveFile;
```
Comment (dkohlbre) [inline lib/host/src/keystone.cpp]: Make sure all the error bailouts also deallocate as needed.

File: lib/host/src/keystone.cpp [hunk 2]
```diff
@@ -118,42 +282,36 @@ keystone_status_t Keystone::destroy()
 
 #define KEYSTONE_ENCLAVE_EDGE_CALL_HOST  11
```
Comment (dkohlbre) [inline lib/host/src/keystone.cpp]: Pretty sure I'm the one that left this here, we should move it out. Optional on this PR.

File: lib/host/src/keystone.cpp [hunk 3]
```diff
@@ -4,111 +4,275 @@
...
+    unsigned long req_pages = 0;
+
+    req_pages += ceil(eapp_sz / PAGE_SIZE);
+    req_pages += ceil(eapp_stack_sz / PAGE_SIZE);
+    req_pages += ceil(rt_sz / PAGE_SIZE);
+    req_pages += ceil(rt_stack_sz / PAGE_SIZE);
+
+    // FIXME: calculate the required number of pages for the page table.
+    // For now, we must allocate at least 1 (top) + 2 (enclave) + 2 (runtime) pages for pg tables
+    req_pages += 15;
```
[the remark this answers is not in the document]
Comment (dayeol) [reply lib/host/src/keystone.cpp]: Ah yeah, this is a problem. If we want to get a precise number of pages that an enclave needs, we need so calculate how many page tables the enclave will use. I just gave 15 extra pages (which is way more than the enclave needs), but this is definitely a bad thing to do here.

File: lib/host/src/keystone.cpp (hunk 1 above)
Comment (dayeol) [reply lib/host/src/keystone.cpp]: Got it.

**Pull request merged**